### ISSUE TITLE
Login: 'Sign In' button in lower left no longer working with proxy config setting root_url grafana >= 9.3.0 

### ISSUE THREAD

**Issue by tsgit** [grafana/grafana#59767]
**What happened**:
Click on "Sign In" in lower left (box with inward arrow) -- get 404 Not Found in response. the  URL is missing subpath
**What you expected to happen**:
be taken to the login screen
**How to reproduce it (as minimally and precisely as possible)**:
set root_url in /etc/grafana/grafana.ini  with a subpath
all other links appear to work, but "Sign In" button is missing the subpath
**Anything else we need to know?**:
the current config worked fine with all grafana 9.2.x and older. this broke in grafana 9.3.0 release
**Environment**:
- Grafana version: >= 9.3
- Data source type & version:
- OS Grafana is installed on: redhat linux
- User OS & Browser: firefox and chrome on redhat linux and fedora
- Grafana plugins: -
- Others: login via the grafana icon in top left still works properly with subpath

**Comment by robezonetv** [comment]
I can confirm it on my environment too:

- Grafana version 9.3.1
- OS: Debian 11.5
- Others: i can confirm above issue, that top login works fine, but left bottom 'Log In' button not use subpath

I found that simmilar issue exists in previous version, but I'm not able to fix it here manualy. ([link](https://github.com/grafana/grafana/pull/26239))

**Comment by zuchka** [comment]
Thanks for creating this issue, @tsgit 

sounds similar to this, would you say?

https://github.com/grafana/grafana/issues/58290#issuecomment-1331330308

**Comment by tsgit** [comment]
PS:  some additional info

`grafana.ini` has
```
root_url = http://<redacted>:8300/system/grafana-dev
```

apache httpd proxy config
```
<Location /system/grafana-dev/>
  Use WebAuth
  ProxyPass http://<redacted>:8300/
  ProxyPassReverse http://<redacted>:8300/
  ProxyPreserveHost On
</Location>
```
the grafana icon in the top left links to 
  `https://<redacted>/system/grafana-dev/login`
 while the "Sign In" box links to 
  `https://<redacted>/d/unoR5g4Wz/pcds-home?orgId=1&forceLogin=true`

**Comment by okimiko** [comment]
I can confirm this issue, too.

- Grafana version: 9.3.0
- Data source type & version: -
- OS Grafana is installed on: docker-image - grafana/grafana:9.3.0
- User OS & Browser: Windows Firefox
- Grafana plugins: -
- Others: Login link is broken

**Comment by tsgit** [comment]
I use the oss version
how to reporduce:
- install https://dl.grafana.com/oss/release/grafana-9.3.1-1.x86_64.rpm
- cp /usr/share/grafana/conf/sample.ini /etc/grafana/grafana.ini
- modify sample.ini
```
diff -u /usr/share/grafana/conf/sample.ini /etc/grafana/grafana.ini
--- /usr/share/grafana/conf/sample.ini  2022-12-01 09:10:46.000000000 -0700
+++ /etc/grafana/grafana.ini    2022-12-08 09:09:09.918310648 -0700
@@ -50,6 +50,7 @@
 # The full public facing url you use in browser, used for redirects and emails
 # If you use reverse proxy and sub path specify full url (with sub path)
 ;root_url = %(protocol)s://%(domain)s:%(http_port)s/
+root_url = https://192.168.1.119/system/grafana
 
 # Serve Grafana from subpath specified in `root_url` setting. By default it is set to `false` for compatibility reasons.
 ;serve_from_sub_path = false
@@ -474,7 +475,7 @@
 #################################### Anonymous Auth ######################
 [auth.anonymous]
 # enable anonymous access
-;enabled = false
+enabled = true
 
 # specify organization name that should be used for unauthenticated users
 ;org_name = Main Org.
```
- systemctl restart grafana-server
-  visit https://192.168.x.x/system/grafana/
![grafana-login](https://user-images.githubusercontent.com/766159/206516933-3abda36e-1f12-4a8b-95ad-45c0d49178a8.png)

the "Sign In" link does not contain the sub_path. 

All other links on that page contain the sub_path and the site is otherwise fully functional. The grafana icon in the upper left points to https://192.168.x.x/system/grafana/login.

The apache httpd conf is 
```
        <Location "/system/grafana/">
          # preserve Host header to avoid cross-origin problems
          ProxyPreserveHost on
          # proxy to grafana
          ProxyPass         http://127.0.0.1:3000/
          ProxyPassReverse  http://127.0.0.1:3000/
          RequestHeader     set "X-Forwarded-Proto" expr=%{REQUEST_SCHEME}
        </Location>
```

**Comment by okimiko** [comment]
@ashharrison90 I'm using the OSS version with the following parameters:

```
[server]
root_url = %(protocol)s://%(domain)s:%(http_port)s/grafana
serve_from_sub_path = true
[auth.anonymous]
enabled = true
```

As far as I know the issue was on all pages (I have no access to my development system at the moment, sorry).

**Comment by okimiko** [comment]
I searched the code a bit: This one changed the behaviour (especially public/app/core/components/NavBar/utils.ts and the removed function "getForcedLoginUrl") and was released with 9.3.x:
https://github.com/grafana/grafana/commit/a8b883b1fa20c1299d5da9bd22d78f5c5586039a#diff-57220af44da9638bd43a9ee29c3ad488359afc40b5c46226c91d139027bde059R37

**Comment by joshhunt** [comment]
Thanks for the extra information. Able to reproduce this clearly when using a reverse proxy such as Apache.

If you're not using a reverse proxy (so just accessing Grafana's built-in HTTP server and `serve_from_sub_path=true`) , it's slightly trickier to notice this because Grafana will automatically redirect `/` to `/subpath`. You can still reproduce this by noticing the `href` on the Sign In `<a />` doesn't contain the subpath

![image](https://user-images.githubusercontent.com/46142/207134049-3b85f9cf-8079-4b9d-8a78-68a076e613dc.png)

**Comment by UlrichThiess** [comment]
Still not working in 9.3.2 on Windows. :-(

**Comment by gkavboy** [comment]
Сonfirm the problem. ALT Linux 8.4
Not work on Grafana >= 9.3.0. Version 9.2.* work well

**Comment by amotl** [comment]
Hi there,

we are able to confirm the problem after upgrading to Grafana 9.3.2 on https://weather.hiveeyes.org/, even after adding the `serve_from_sub_path = true` configuration setting.

@zuchka: Can you also add the https://github.com/grafana/grafana/labels/type%2Fregression label here, and maybe vote for adding it to the next milestone? Thank you in advance!

With kind regards,
Andreas.

**Comment by amotl** [comment]
@okimiko: Thank you for looking into the details.

> I searched the code a bit: This one changed the behaviour (especially `public/app/core/components/NavBar/utils.ts` and the removed function `getForcedLoginUrl`) and was released with 9.3.x: a8b883b1.

Let me reference the corresponding issues / pull requests. here.

### References
- GH-56403
- GH-58290

**Comment by tsgit** [comment]
same issue with release 9.3.4

**Comment by virtualdreams** [comment]
Same issue with release 9.3.6.

**Comment by gathar** [comment]
Same issue in Grafana Enterprise (9.3.4).

**Comment by ghassencherni** [comment]
Not fixed

**Comment by Robertof** [comment]
It looks like this fix hasn't made it in 9.4.0-beta1 for a day and there is still no version after that. For anyone having this issue and using `nginx` as a reverse proxy for Grafana, this is the simplest fix I found:

```nginx
location / {
  if ($arg_forceLogin = true) {
    rewrite ^ /grafana$request_uri;
  }
}
```

(do note that if you have any other web application which uses the `forceLogin` query parameter though, this will break. But if not, it works like a charm :-))

**Comment by diraol** [comment]
> It looks like this fix hasn't made it in 9.4.0-beta1 for a day and there is still no version after that. For anyone having this issue and using `nginx` as a reverse proxy for Grafana, this is the simplest fix I found:
> 
> ```nginx
> location / {
>   if ($arg_forceLogin = true) {
>     rewrite ^ /grafana$request_uri;
>   }
> }
> ```
> 
> (do note that if you have any other web application which uses the `forceLogin` query parameter though, this will break. But if not, it works like a charm :-))

You may also try using the `referer` request header to check where is the request coming from. (`$http_referer`).

```nginx
location / {
  if ($arg_forceLogin = true && $http_referer =~ "https?://<domain>/grafana") {
    rewrite ^ /grafana$request_uri;
  }
}
```

(Disclaimer: I haven't tested it)

**Comment by UlrichThiess** [comment]
Works with 9.5.0-102803-pre
